### PR TITLE
Make bootstrap client cert loading part of rotation

### LINKED ISSUE #68686: Kubelet does not launch static pods while waiting for bootstrapping
The kubelet currently blocks trying to attempt to get a bootstrap client before it starts its worker loops for running static pods, which means a static pod that provides functionality on a node that bootstrapping may need (talking to an HSM, calling out externally, providing a plugin for bootstrapping, or even running a control plane pod) is blocked.  There is no actual functionality in the Kubelet that depends on this (waiting for bootstrap has no use case value), it's just a historical legacy of how the code evolved.  As people start to leverage bootstrapping and rotation, it becomes more obvious how this doesn't work in practice.

In the long run we want bootstrap credentials to be pluggable and extensible so that an HSM or other trust store on the machine can deal with the problem.  Having secret key material for bootstrapping stored on disk that allows you to join a cluster is undesirable, and having a callout to make that flow less error prone is very desirable for both cloud and metal provisioned environments.

In addition, the bootstrapping and client cert rotation flows are slightly duplicated today - using the client cert to rotate itself means that when a client cert expires (very possible with quicker rotations if the api goes down for a period of time) the bootstrap credential is consulted - but that could be arbitrarily far in the future and means we tend to fail in those cases (bootstrap credential expired).  Since the long term desire is for bootstrap to be out of process, it would be better for us to continually go back to bootstrap on cert expiration rather than retrying our flow.

For 1.13, we should fix the kubelet to not pause when bootstrapping and instead delegate to the rotation flow to provide that credential.  This has the side effect of reducing duplicated code for procuring CSRs between the two code paths, and hiding the details of client cert acquisition inside of the cert manager so that future changes can allow call outs.

@awly @mikedanese as discussed this week, @kubernetes/sig-auth-misc / @kubernetes/sig-node-bugs, @liggitt because you and discussed this, @aaronlevy because we wanted to push allow static pods for bootstrap control planes in bootkube not to block when using bootstrapping kubelets..

### PR DESCRIPTION
**Pull request by smarterclayton**
Ensure that bootstrap+client-cert rotation in the Kubelet can:

1. happen in the background so that static pods are launched immediately
2. collapse down to a single CSR request codepath
3. reorganize the code to allow future improvements to fetching bootstrap creds

Reorganize how the kubelet client config is determined. If rotation is
off, simplify the code path. If rotation is on, load the config
from disk, and then pass that into the cert manager. The cert manager
creates a client each time it tries to request a new cert.

Preserves existing behavior where:

1. bootstrap kubeconfig is used if the current kubeconfig is invalid/expired
2. we create the kubeconfig file based on the bootstrap kubeconfig, pointing to
   the location that new client certs will be placed
3. the newest client cert is used once it has been loaded

reverted in https://github.com/kubernetes/kubernetes/pull/71173, original release note was:
```
When a kubelet is using --bootstrap-kubeconfig and certificate rotation, it no longer waits for bootstrap to succeed before launching static pods.
```

```release-note
NONE
```

Fixes #68686

### CODE REVIEW

**Comment by smarterclayton** [comment]
/assign @awly 

@mikedanese @liggitt as discussed, @aaronlevy will allow bootkube and other self-hosting mechanisms to rely on bootstrapped master nodes and use static pods.

**Comment by smarterclayton** [comment]
I reverted #66056 in this - perhaps we could instead do a more aggressive dial timeout for bootstrap credentials (now that we always do both in the background)?

**Comment by smarterclayton** [comment]
Testing this here and in openshift/origin#21274 so I have a better baseline across different environments

**Comment by smarterclayton** [comment]
There are four core scenarios to test:

1. rotation on, bootstrap on
  1. no client certs, --kubeconfig file doesn't exist -> kubeconfig is created pointed at destination of client certs, cert is refreshed
  2. client certs exist, --kubeconfig file doesn't exist -> kubeconfig is created pointing at client certs, cert is not refreshed
  3. client certs exist, --kubeconfig exists, certs are valid -> cert is not refreshed until next interval
  3. client certs exist, --kubeconfig exists, certs are expired -> static pod starts and cert is refreshed
2. rotation off, bootstrap on
  1. bootstrap cert is expired -> rebootstrap
  2. bootstrap file doesn't exist -> exit with error
  3. regular kubeconfig exists but is invalid -> exit with error
3. rotation off, bootstrap off
  1. kubeconfig doesn't exist -> exit with error
  2. kubeconfig exists but is invalid/expired -> static pods start, errors are logged
4. rotation on, bootstrap off

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -771,6 +731,74 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
...
+			return nil, nil, err
+		}
+
+		newClientFn := func(current *tls.Certificate) (certificatesclient.CertificateSigningRequestInterface, error) {
+			// If we have a valid certificate, use that to fetch CSRs. Otherwise use the bootstrap
+			// credentials.
+			// XXX: When an external bootstrap source is available, it should be possible to always use that source
+			//      to retrieve new credentials.
+			config := certConfig
+			if current != nil {
```
Comment (awly) [inline cmd/kubelet/app/server.go]: IIUC, if certificate manager has an expired cert and fails to rotate it, `Current()` will return that expired cert.
Can we check that `current` is still valid here and fall back to certConfig if it's not?

File: pkg/kubelet/certificate/bootstrap/bootstrap.go [hunk 2]
```diff
@@ -43,6 +38,55 @@ import (
...
+func LoadClientConfig(kubeconfigPath string, bootstrapPath string, certDir string) (certConfig, userConfig *restclient.Config, err error) {
+	if len(bootstrapPath) == 0 {
+		clientConfig, err := loadRESTClientConfig(kubeconfigPath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to load kubeconfig: %v", err)
+		}
+		return clientConfig, clientConfig, nil
+	}
+
+	store, err := certificate.NewFileStore("kubelet-client", certDir, certDir, "", "")
```
Comment (awly) [inline pkg/kubelet/certificate/bootstrap/bootstrap.go]: Is this only used to generate `pemPath`?
moveit down to where it's used.

File: pkg/kubelet/certificate/bootstrap/bootstrap.go [hunk 3]
```diff
@@ -43,6 +38,55 @@ import (
 
 const tmpPrivateKeyFile = "kubelet-client.key.tmp"
 
+// LoadClientConfig tries to load the appropriate client config for retrieving certs and for use by users.
+// If bootstrapPath is empty, only kubeconfigPath is checked. If bootstrap path is set and the contents
+// of kubeconfigPath are valid, both certConfig and userConfig will point to that file. Otherwise the
+// kubeconfigPath on disk is populated based on bootstrapPath but pointing to the location of the client cert
+// in certDir. This preserves the historical behavior of bootstrapping where on subsequent restarts the
+// most recent client cert is used to request new client certs instead of the initial token.
+func LoadClientConfig(kubeconfigPath string, bootstrapPath string, certDir string) (certConfig, userConfig *restclient.Config, err error) {
```
Comment (awly) [inline pkg/kubelet/certificate/bootstrap/bootstrap.go]: bootstrapPath, certDir string

File: cmd/kubelet/app/server.go [hunk 4]
```diff
@@ -771,6 +731,74 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
...
+			return nil, nil, err
+		}
+
+		glog.V(2).Info("Starting client certificate rotation.")
+		clientCertificateManager.Start()
+
+		return &transportConfig, closeAllConns, nil
+	}
+
+	if len(s.BootstrapKubeconfig) > 0 {
```
Comment (awly) [inline cmd/kubelet/app/server.go]: Should this get called before starting the cert managed above?

**Review by awly — COMMENTED** [review]
Re #66056: it should be fine to remove if we add aggressive timeout and _fast_ retries for initial bootstrap.

For context: it was needed on GCE because network programming can lag behind and bootstrapping would get trapped in a very long timeout, even after master IP becomes reachable from Node. cc @mikedanese

File: cmd/kubelet/app/server.go (hunk 4 above)
Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: this is in the unmanaged path.  In the managed path bootstrap is managed by the manager.

File: pkg/kubelet/certificate/bootstrap/bootstrap.go (hunk 2 above)
Comment (smarterclayton) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: it's high in the function to fail earlier on local config

File: cmd/kubelet/app/server.go (hunk 1 above)
Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: Current() returns nil if the cert is expired.

Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: Or at least it should, that was the intended behavior a while back.  The transport rotation loop checks for the transition from valid -> invalid and uses that as the "die if this persists"

Comment (awly) [reply cmd/kubelet/app/server.go]: The cached cert is only updated on successful rotation: http://go/gh/kubernetes/kubernetes/blob/3c493bc95a825f0f6eb38ef78a19811f204053ec/staging/src/k8s.io/client-go/util/certificate/certificate_manager.go#L396
And `Current()` just returns the cached cert without checking expiry http://go/gh/kubernetes/kubernetes/blob/3c493bc95a825f0f6eb38ef78a19811f204053ec/staging/src/k8s.io/client-go/util/certificate/certificate_manager.go#L204-L208

Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: Ah, I added the filtering to getCurrentCertificateOrBootstrap.  There's very little point in the cert manager handing me back an expired cert - can you think of a case where that makes sense?  If not, we may want to change the signature and verify callers get the right thing (callers should have to deal with nil in general).  Alternatively, we make every caller filter for a valid cert.

Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: ok, I agree.  I'll make that change and add tests

**Comment by smarterclayton** [comment]
still looking at making the waitForServer path able to have a more aggressive timeout and dial.

**Comment by smarterclayton** [comment]
Ok, I *think* the last commit addresses the issue, but still need to test.  Using a custom dial function on a rest client config busts the cache, so we always get a new transport.  I refactored wait (long time coming anyway) to give us sequence numbers.  We gradually increase the time we wait until we hit a limit, only retrying on timeout.  Feels a bit awkward to treat timeout special, but is close enough for government work.

Only problem with the commit as is that it fills the TLS cache with infinite entries if anyone used LoadClientCert too much.  So I'll change this before we merge to bypass the TLS cache completely, or talk with @liggitt about whether we need to make the cache an LRU or bypass the cache with a custom dial function.

**Comment by liggitt** [comment]
> Only problem with the commit as is that it fills the TLS cache with infinite entries if anyone used LoadClientCert too much. So I'll change this before we merge to bypass the TLS cache completely, or talk with @liggitt about whether we need to make the cache an LRU or bypass the cache with a custom dial function.

Custom dial functions must produce unique cache keys. 

If you want to guarantee no stuck tcp connections after timeout, I think having an option to disable keep alive when constructing the transport would be better

File: pkg/kubelet/certificate/bootstrap/bootstrap.go [hunk 5]
```diff
@@ -98,23 +139,47 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
...
+	interval := wait.Backoff{
+		Duration: time.Second,
+		Factor:   1.1,
+		Steps:    20,
+		Jitter:   0.2,
+		Cap:      30 * time.Second,
 	}
+	for interval.Steps > 0 {
+		// creating a new client with a new Dial function ensures we get a new cache entry
+		// TODO: it also leaks cache entries, which is a broader problem
```
Comment (mikedanese) [inline pkg/kubelet/certificate/bootstrap/bootstrap.go]: https://github.com/kubernetes/kubernetes/pull/58677 would help here.

We need to make sure transports are also cleaned up. I don't think that will happen properly when bootstrapClient is GC'd:

https://github.com/golang/go/issues/24739

Comment (smarterclayton) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: It won't, which is why Jordan suggested a "bypass reuse" option.  I don't think the TLS cache holding things forever is a reasonable long term path, but I could be persuaded it's a short term avoidable problem.

I could just create a separate transport here and bypass the cache completely, which may just be the simplest option.

File: pkg/kubelet/certificate/bootstrap/bootstrap.go [hunk 6]
```diff
@@ -98,23 +140,64 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
...
+		t := interval.Step()
+		timeoutClientConfig.Transport = &http.Transport{
+			TLSClientConfig:   tlsConfig,
+			DisableKeepAlives: true,
+			DialContext:       (&net.Dialer{Timeout: t}).DialContext,
+		}
+
+		bootstrapClient, err := certificates.NewForConfig(timeoutClientConfig)
+		if err != nil {
+			if t, ok := err.(timeoutError); ok && t.Timeout() {
```
Comment (awly) [inline pkg/kubelet/certificate/bootstrap/bootstrap.go]: Why would creating a new certs client timeout?

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go [hunk 7]
```diff
@@ -177,6 +177,32 @@ type Backoff struct {
 	Factor   float64       // Duration is multiplied by factor each iteration
 	Jitter   float64       // The amount of jitter applied each iteration
 	Steps    int           // Exit with error after this many steps
+	Cap      time.Duration // The maximum amount of time to wait if set
+}
+
+// Step returns the next interval in the exponential backoff.
+func (b *Backoff) Step() time.Duration {
```
Comment (awly) [inline staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go]: I know it's annoying, but can you move this refactor to separate PR?

File: pkg/kubelet/certificate/bootstrap/bootstrap.go (hunk 6 above)
Comment (smarterclayton) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: This is the first api call made to join a node to a cluster, it races with network programming on GCE, GCE has bugs where ACKs disappear into the ether.  See the referenced PR

File: pkg/kubelet/certificate/bootstrap/bootstrap.go [hunk 8]
```diff
@@ -98,23 +140,64 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
...
+	timeoutClientConfig := bootstrapClientConfig.DeepCopy()
+	timeoutClientConfig.TLSClientConfig = transport.TLSClientConfig{}
+
+	for interval.Steps > 0 {
+		// use a custom transport that disables keep alives
+		t := interval.Step()
+		timeoutClientConfig.Transport = &http.Transport{
+			TLSClientConfig:   tlsConfig,
+			DisableKeepAlives: true,
+			DialContext:       (&net.Dialer{Timeout: t}).DialContext,
```
[the remark this answers is not in the document]
Comment (smarterclayton) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: No, it’s not thread safe to change a dialect under the covers.  Default dialers aren’t expensive.

File: pkg/kubelet/certificate/bootstrap/bootstrap.go [hunk 9]
```diff
@@ -43,11 +41,60 @@ import (
...
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to build bootstrap cert store")
+	}
+
+	ok, err := verifyBootstrapClientConfig(kubeconfigPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// use the current client config
```
[the remark this answers is not in the document]
Comment (smarterclayton) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: Should we?  Probably.  But this is the current behavior.  And after five minutes without a rotation the kubelet restarts.  So preserving legacy behavior.

File: staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go (hunk 7 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go]: I will once we tease apart the retry design.

File: pkg/kubelet/certificate/bootstrap/bootstrap.go (hunk 6 above)
Comment (awly) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: But isn't `csr.RequestNodeCertificate` below the API call? Why would `certificates.NewForConfig` result in a request?

File: pkg/kubelet/certificate/bootstrap/bootstrap.go [hunk 10]
```diff
@@ -98,23 +140,64 @@ func LoadClientCert(kubeconfigPath string, bootstrapPath string, certDir string,
...
-		klog.V(2).Infof("failed cleaning up private key file %q: %v", privKeyPath, err)
+	timeoutClientConfig := bootstrapClientConfig.DeepCopy()
+	timeoutClientConfig.TLSClientConfig = transport.TLSClientConfig{}
+
+	for interval.Steps > 0 {
+		// use a custom transport that disables keep alives
+		t := interval.Step()
+		timeoutClientConfig.Transport = &http.Transport{
+			TLSClientConfig:   tlsConfig,
+			DisableKeepAlives: true,
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: See here https://github.com/kubernetes/kubernetes/pull/69890#discussion_r232780288 for justification. Maybe add a comment?

**Comment by smarterclayton** [comment]
Add a test to verify we don't regress the core client behavior (use the bootstrap client, then the retrieved client).

File: pkg/kubelet/certificate/bootstrap/bootstrap.go (hunk 6 above)
Comment (smarterclayton) [reply pkg/kubelet/certificate/bootstrap/bootstrap.go]: Moved out of this PR so as not to block this if we can't get review time by tomorrow.

**Comment by smarterclayton** [comment]
@awly this is back to effectively the original PR you reviewed.  Was your previous review sufficient or were you looking for more?

**Review by mikedanese — COMMENTED** [review]
I'm +1 on getting this in to CI and seeing how it goes.

/lgtm
/approve

File: cmd/kubelet/app/server.go [hunk 11]
```diff
@@ -610,28 +584,15 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
...
 		heartbeatClientConfig.QPS = float32(-1)
-		heartbeatClient, err = clientset.NewForConfig(&heartbeatClientConfig)
+		kubeDeps.HeartbeatClient, err = clientset.NewForConfig(&heartbeatClientConfig)
 		if err != nil {
-			klog.Warningf("Failed to create API Server client for heartbeat: %v", err)
+			return fmt.Errorf("failed to initialize kubelet heartbeat client: %v", err)
 		}
 
-		// csiClient works with CRDs that support json only
-		clientConfig.ContentType = "application/json"
```
[the remark this answers is not in the document]
Comment (liggitt) [reply cmd/kubelet/app/server.go]: pre-merge tests wouldn't pick up failures here, since it's alpha. need to verify this still works before merge

**Comment by liggitt** [comment]
/hold
for the csiClient question

File: cmd/kubelet/app/server.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: I'll restore with comments linking to the issue.  I'm surprised at renegotiation not working on watch, I thought we had tests for it.

File: cmd/kubelet/app/server.go [hunk 12]
```diff
@@ -771,6 +732,81 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
...
+}
+
+// buildClientCertificateManager creates a certificate manager that will use certConfig to request a client certificate
+// if no certificate is available, or the most recent clientConfig (which is assumed to point to the cert that the manager will
+// write out).
+func buildClientCertificateManager(certConfig, clientConfig *restclient.Config, certDir string, nodeName types.NodeName) (certificate.Manager, error) {
+	newClientFn := func(current *tls.Certificate) (certificatesclient.CertificateSigningRequestInterface, error) {
+		// If we have a valid certificate, use that to fetch CSRs. Otherwise use the bootstrap
+		// credentials.
+		// XXX: When an external bootstrap source is available, it should be possible to always use that source
```
[the remark this answers is not in the document]
Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: picked up another team's convention and used it here.

File: cmd/kubelet/app/server.go [hunk 13]
```diff
@@ -537,66 +538,39 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
...
-	}
-
 	// if in standalone mode, indicate as much by setting all clients to nil
 	if standaloneMode {
 		kubeDeps.KubeClient = nil
 		kubeDeps.DynamicKubeClient = nil
 		kubeDeps.EventClient = nil
 		kubeDeps.HeartbeatClient = nil
 		klog.Warningf("standalone mode, no API client")
+
```
[the remark this answers is not in the document]
Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: intentional to make the blocks easier to read.  I may make this a switch instead.

**Comment by smarterclayton** [comment]
Comments addressed

**Comment by smarterclayton** [comment]
This was reverted in #71173 due to breaking kubeadm's requirement that `--kubeconfig` be provided on masters as a high powered credential and cause a rotation of the cert.  Restored in #71174 with coverage for kubeadm's flow.

**Pull request merged**